compiler/sem/semdata.nim

```diff
@@ -91,39 +91,339 @@ type
+
+    features*: set[Feature]    ## compiler feature flags, can be varied via
+                               ## pragmas, so per module; maybe that shouldn't
+                               ## be a thing?
+    
+    # only used in `importer`
+    importModuleMap*: Table[int, int] ## (module.id, module.id)
+    exportIndirections*: HashSet[(int, int)] ## (module.id, symbol.id)
+    recursiveDep*: seq[tuple[importer, importee: string]]
+      ## used to detect recursive `importer` issues populated by importer and
+      ## used by `lookups`
+    # -------------------------------------------------------------------------
+    # end: module env; module lifetime
+    # -------------------------------------------------------------------------
+
+
+    # -------------------------------------------------------------------------
+    # start: caches/common defs, likely whole program, per modules sandboxing?
+    # -------------------------------------------------------------------------
+    
+    # ident cache used everywhere; sandboxed to avoid pollution?
+    cache*: IdentCache
+
+    # module graph, access back to the whole compilation
+    graph*: ModuleGraph
+
+    # type lookups
+    voidType*: PType           ## for typeof(stmt)
+    intTypeCache*: array[-5..32, PType] ## cache some common integer types
+                                        ## to avoid type allocations
+    nilTypeCache*: PType
+
+    # symbols/dispatch
+    converters*: seq[PSym]     ## track current converters and then look them
+                               ## up during dispatch (`sigmatch`)
+    pureEnumFields*: TStrTable ## pure enum fields that can be used unambiguously
+
+    # term rewriting/hlo
+    patterns*: seq[PSym]       ## sequence of pattern matchers used for term
+                               ## rewriting and also `hlo`, where highlevel
+                               ## optimizations are effectively pattern matches
+    
+    # pragmas and codegen ref
+    userPragmas*: TStrTable
+    libs*: seq[PLib]           ## all libs used by this module, mostly setup
+                               ## pragmas.
+
+    # hacks used for lookups
+    isAmbiguous*: bool # little hack <-- it never is "little"
+    # -------------------------------------------------------------------------
+    # end: caches/common defs, likely whole program, per modules sandboxing?
+    # -------------------------------------------------------------------------
+
+
+    # -------------------------------------------------------------------------
+    # start: not entirely clear why, function pionters for certain sem calls?
+    # -------------------------------------------------------------------------
     semConstExpr*: proc (c: PContext, n: PNode): PNode {.nimcall.} # for the pragmas
     semExpr*: proc (c: PContext, n: PNode, flags: TExprFlags = {}): PNode {.nimcall.}
     semTryExpr*: proc (c: PContext, n: PNode, flags: TExprFlags = {}): PNode {.nimcall.}
```

Comment:
I believe this is done as a way to make various module (e.g. `pragmas`) importable instead of having to use includes or cyclic imports

Reply:
Ah thanks, I knew this fact at some point and then forgot while doing this work.

Aside: English hilarity, I first read the "I'm" in "importable" as it is in the word "impossible". Took a second to parse your comment as intended. 😄